[WIP] Event editor

melody-rs opened the pull request:
**Connections**
N/A

**Description**
This reimplements (part of) the event editor, as well as the picker modals! (Switch, Variable, Graphic, etc) I added some temporary test modals into the item and actor editor to show off that they work.
There isn't any event command editor yet though- That's reserved for another PR. I'm still working out the details of the event command editor. Same goes for Undo/Redo as well.

**Testing**
- Open a project
- Navigate to any map
- Open an event
  - Edit switches
  - Switch pages
  - Add and delete pages

<!-- 
Thanks for filing a pull request! The codeowners file will automatically request reviews.

After you get a review and have addressed any comments, please explicitly re-request a review from the
person(s) who reviewed your changes. This will make sure it gets re-added to their review queue - you're not bothering us!
-->

**Checklist**

- [x] Run `cargo fmt`.
- [ ] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown -Z build-std=std,panic_abort`
- [x] Run `cargo build --release`
- [x] If applicable, run `trunk build --release`

melody-rs committed:
forget outputstream, making audio send+sync

melody-rs committed:
Port over modals from old branch

melody-rs committed:
add basic event editor window

melody-rs committed:
Fix modal id conflict

melody-rs committed:
Use unified modal

melody-rs committed:
Replace import hack

melody-rs committed:
Add most of the event ui

melody-rs committed:
start implementing the graphic picker

melody-rs committed:
Add picker window

melody-rs committed:
Display graphic picker options

melody-rs committed:
Add sorta working tilepicker

melody-rs committed:
More picker stuff

melody-rs committed:
More picker things

melody-rs committed:
Hue & Opacity editing

melody-rs committed:
Fix some display issues

melody-rs committed:
adjust layout

melody-rs committed:
Show stripes

melody-rs committed:
Add search and fix stripes

melody-rs committed:
Actually save the graphic

melody-rs committed:
Insert events

melody-rs committed:
Minor fixes

melody-rs committed:
Validate texture sizes

melody-rs committed:
Fix fmt errors

melody-rs committed:
Merge remote-tracking branch 'upstream/dev' into event-editor

melody-rs committed:
Make preview sprite loading more fallible

melody-rs committed:
Grey out invalid sprites

melody-rs committed:
Fix preview sprite handling

melody-rs committed:
Properly handle sprite scrolling

melody-rs committed:
Load from RTP on native too

melody-rs committed:
Properly load RTPs this time

melody-rs committed:
Don't enable debug symbols in release

melody-rs committed:
Remove all nightly features

melody-rs committed:
Update readme

melody-rs committed:
Clean up soundtab

melody-rs committed:
cache sound tab results

melody-rs committed:
Cache filters

melody-rs committed:
Use rows for graphic picker

melody-rs committed:
Re-add stripes to graphic picker

melody-rs committed:
Remove first open

melody-rs committed:
Very destructive commit

Not very happy with this because it clones things, which is a waste of memory

melody-rs committed:
Revert "Very destructive commit"

This reverts commit 471b014c55908fe9de0a362142e265f005428aa3.

white-axe commented on crates/ui/src/windows/event_edit.rs:
```diff
@@ -22,55 +22,252 @@
+            .show(ctx, |ui| {
+                let id_source = self.id();
+                let previous_page = self.selected_page;
+                egui::TopBottomPanel::top(id_source.with("top_panel")).show_inside(ui, |ui| {
+                    ui.horizontal(|ui| {
+                        ui.label("Name: ");
+                        ui.text_edit_singleline(&mut self.event.name);
+                    });
+                    ui.horizontal(|ui| {
+                        ui.label("Page: ");
+                        for i in 0..self.event.pages.len() {
+                            ui.selectable_value(&mut self.selected_page, i, format!("{}", i + 1));
+                        }
+
+                        if ui
+                            .button(egui::RichText::new("Add").color(egui::Color32::LIGHT_GREEN))
+                            .clicked()
+                        {
+                            self.event.pages.push(rpg::EventPage::default());
+                            self.selected_page = self.event.pages.len() - 1;
+                        }
+
+                        let button = egui::Button::new(
+                            egui::RichText::new("Delete").color(egui::Color32::LIGHT_RED),
+                        );
+                        if ui.add_enabled(self.event.pages.len() > 1, button).clicked() {
+                            self.event.pages.remove(self.selected_page);
+                            self.selected_page = self.selected_page.saturating_sub(1);
+                        }
+                        if ui.button(egui::RichText::new("Clear")).clicked() {
+                            self.event.pages[self.selected_page] = rpg::EventPage::default();
+                        }
+                    });
+                });
```
Can you add a little bit of spacing above and below this top panel in the event editor window? It looks a bit odd when the bottom edges of the buttons in the top panel are flush with the separator. Same goes for the top panel in the graphic picker, although that one only needs extra spacing above.

white-axe commented on crates/components/src/lib.rs:
```diff
@@ -84,6 +83,36 @@ impl<'e, T: ToString + PartialEq + strum::IntoEnumIterator> egui::Widget for Enu
+    }
+}
+
+impl<'e, T: ToString + PartialEq + strum::IntoEnumIterator> egui::Widget for EnumRadioList<'e, T> {
+    fn ui(self, ui: &mut egui::Ui) -> egui::Response {
+        let mut changed = false;
+        let mut response = ui
+            .vertical(|ui| {
+                for variant in T::iter() {
+                    let text = variant.to_string();
+                    if ui.radio_value(self.current_value, variant, text).changed() {
+                        changed = true;
+                    }
+                }
+            })
+            .response;
+        if changed {
+            response.mark_changed();
+        }
+        response
+    }
+}
```
Why does the left "Options" column of the event editor window expand when the event editor window is resized horizontally while the right "Trigger" column stays the same size? I think the right column should also expand when the window is resized for consistency. One way to fix this would be to use `ui.with_cross_justify` (defined in crates/components/src/ui_ext.rs) inside of the `EnumRadioList` widget.

```suggestion
impl<'e, T: ToString + PartialEq + strum::IntoEnumIterator> egui::Widget for EnumRadioList<'e, T> {
    fn ui(self, ui: &mut egui::Ui) -> egui::Response {
        let mut changed = false;
        let mut response = ui
            .vertical(|ui| {
                ui.with_cross_justify(|ui| {
                    for variant in T::iter() {
                        let text = variant.to_string();
                        if ui.radio_value(self.current_value, variant, text).changed() {
                            changed = true;
                        }
                    }
                });
            })
            .response;
        if changed {
            response.mark_changed();
        }
        response
    }
}
```

melody-rs committed:
Only load directory entries on modal opening

melody-rs committed:
Pad out top bar of event editor & graphic picker

melody-rs committed:
Use cross justify to expand the trigger window

melody-rs committed:
Fmt

melody-rs committed:
mem::take data from update state so we can use it in database view

melody-rs committed:
Add menu use se modal to item editor

melody-rs committed:
Use get_mut instead of borrow() in Data::save()

melody-rs committed:
Clean up sound picker modal a little

melody-rs committed:
Only load tilepicker when selecting tiles

melody-rs committed:
work on the graphic picker a bit

melody-rs committed:
Make Modal::reset actually useful

melody-rs committed:
Add graphic picker to item editor

melody-rs committed:
Fill out graphic picker more

melody-rs committed:
Make modals take data by value (might revert this)

melody-rs committed:
Remove ResetData
Was annoying and probably not needed

melody-rs committed:
Move current graphic picker into "basic"

melody-rs committed:
Add hue picker

melody-rs committed:
Deduplicate graphic picker behaviour

melody-rs committed:
Move event graphic picker into graphic_picker/

melody-rs committed:
Deduplicate more things

melody-rs committed:
Hack in the enemy battler picker

melody-rs committed:
add hacky actor graphic picker

white-axe commented on crates/ui/src/windows/event_edit.rs:
```diff
@@ -22,55 +22,256 @@
+                );
+            });
+
+        if needs_save {
+            self.event.extra_data.modified.set(true);
+            let mut map = update_state.data.get_map(self.map_id);
+            map.events.insert(self.event.id, self.event.clone()); // don't like the extra clone, but it's necessary
+        }
```
When saving a modified event, the new event that gets inserted into `map.events` needs to have its position updated to account for any changes to the position to the event made in between when the event editor window was opened and when the event was saved.

```suggestion
        if needs_save {
            self.event.extra_data.modified.set(true);
            let mut map = update_state.data.get_map(self.map_id);
            if let Some(old_event) = map.events.get(self.event.id) {
                self.event.x = old_event.x;
                self.event.y = old_event.y;
            }
            map.events.insert(self.event.id, self.event.clone()); // don't like the extra clone, but it's necessary
        }
```

Although maybe it would be better for the event editor to just get a mutable reference to the event it's editing from `update_state` every frame.

melody-rs committed:
Directly* edit events

melody-rs committed:
Remove Window::name

It was just technical debt, back from when Window deduplication was by Window::name instead of Window::id

white-axe commented on crates/ui/src/windows/common_event_edit.rs:
Please fix the compiler warnings in this file

white-axe commented on crates/audio/src/lib.rs:
The changes to the audio handler code seem to have broken the sound test in native builds. I got this panic message when playing an audio file:
```
The application panicked (crashed).
Message:  assertion failed: from >= 1
Location: /home/h/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rodio-0.17.3/src/conversions/sample_rate.rs:53

Luminol version: ec84e01c

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
                                ⋮ 9 frames hidden ⋮                               
  10: rodio::conversions::sample_rate::SampleRateConverter<I>::new::hc8a02c78f36bafb5
      at /home/h/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rodio-0.17.3/src/conversions/sample_rate.rs:<unknown line>
  11: rodio::source::uniform::UniformSourceIterator<I,D>::bootstrap::h4de1d791a863b042
      at /home/h/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rodio-0.17.3/src/source/uniform.rs:66
        64 │             n: frame_len,
        65 │         };
        66 >         let input = SampleRateConverter::new(
        67 │             input,
        68 │             cpal::SampleRate(from_sample_rate),
  12: <rodio::source::uniform::UniformSourceIterator<I,D> as core::iter::traits::iterator::Iterator>::next::h3926ea418d4e9f07
      at /home/h/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rodio-0.17.3/src/source/uniform.rs:102
       100 │ 
       101 │         let mut input =
       102 >             UniformSourceIterator::bootstrap(input, self.target_channels, self.target_sample_rate);
       103 │ 
       104 │         let value = input.next();
                                 ⋮ 1 frame hidden ⋮                               
  14: rodio::dynamic_mixer::DynamicMixer<S>::sum_current_sources::habfbff1ef0182ea2
      at /home/h/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rodio-0.17.3/src/dynamic_mixer.rs:170
       168 │ 
       169 │         for mut source in self.current_sources.drain(..) {
       170 >             if let Some(value) = source.next() {
       171 │                 sum = sum.saturating_add(value);
       172 │                 self.still_current.push(source);
  15: <rodio::dynamic_mixer::DynamicMixer<S> as core::iter::traits::iterator::Iterator>::next::h30621e511426ba97
      at /home/h/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rodio-0.17.3/src/dynamic_mixer.rs:125
       123 │         self.sample_count += 1;
       124 │ 
       125 >         let sum = self.sum_current_sources();
       126 │ 
       127 │         if self.current_sources.is_empty() {
  16: <cpal::platform::platform_impl::Device as rodio::stream::CpalDeviceExt>::new_output_stream_with_format::{{closure}}::{{closure}}::hebe21bf0f3c95ca2
      at /home/h/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rodio-0.17.3/src/stream.rs:219
       217 │                 move |data, _| {
       218 │                     data.iter_mut()
       219 >                         .for_each(|d| *d = mixer_rx.next().unwrap_or(0f32))
       220 │                 },
       221 │                 error_callback,
                                 ⋮ 1 frame hidden ⋮                               
  18: <cpal::platform::platform_impl::Device as rodio::stream::CpalDeviceExt>::new_output_stream_with_format::{{closure}}::h5892f6f263315f67
      at /home/h/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rodio-0.17.3/src/stream.rs:218
       216 │                 &format.config(),
       217 │                 move |data, _| {
       218 >                     data.iter_mut()
       219 │                         .for_each(|d| *d = mixer_rx.next().unwrap_or(0f32))
       220 │                 },
  19: cpal::traits::DeviceTrait::build_output_stream::{{closure}}::hbce3d060dae29ecf
      at /home/h/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cpal-0.15.3/src/traits.rs:169
       167 │             T::FORMAT,
       168 │             move |data, info| {
       169 >                 data_callback(
       170 │                     data.as_slice_mut()
       171 │                         .expect("host supplied incorrect sample type"),
  20: cpal::host::alsa::process_output::h9b9b8e66a87cac90
      at /home/h/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cpal-0.15.3/src/host/alsa/mod.rs:832
       830 │         let timestamp = crate::OutputStreamTimestamp { callback, playback };
       831 │         let info = crate::OutputCallbackInfo { timestamp };
       832 >         data_callback(&mut data, &info);
       833 │     }
       834 │     loop {
  21: cpal::host::alsa::output_stream_worker::h5ac2f4ccb26949f2
      at /home/h/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cpal-0.15.3/src/host/alsa/mod.rs:664
       662 │                     "expected output stream, but polling descriptors indicated input",
       663 │                 );
       664 >                 if let Err(err) = process_output(
       665 │                     stream,
       666 │                     &mut ctxt.buffer,
  22: cpal::host::alsa::Stream::new_output::{{closure}}::h77d65ced486cad65
      at /home/h/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cpal-0.15.3/src/host/alsa/mod.rs:961
       959 │             .name("cpal_alsa_out".to_owned())
       960 │             .spawn(move || {
       961 >                 output_stream_worker(
       962 │                     rx,
       963 │                     &stream,
                                ⋮ 11 frames hidden ⋮                              
```

white-axe commented on crates/components/src/sound_tab.rs:
```diff
@@ -22,57 +22,70 @@
     pub fn new(
         filesystem: &impl luminol_filesystem::FileSystem,
         source: luminol_audio::Source,
+        audio_file: luminol_data::rpg::AudioFile,
     ) -> Self {
-        let folder_children = filesystem.read_dir(format!("Audio/{source}")).unwrap();
+        let mut folder_children = filesystem.read_dir(format!("Audio/{source}")).unwrap();
+        folder_children.sort_unstable_by(|a, b| a.file_name().cmp(b.file_name()));
         Self {
             source,
-            volume: 100,
-            pitch: 100,
-            selected_track: String::new(),
+            audio_file,
+
+            filtered_children: folder_children.clone(),
+            search_text: String::new(),
             folder_children,
         }
     }
```
The volume and pitch should default to 100 like before. The volume defaults to 0 and the pitch defaults to 50 now.

melody-rs committed:
Fix AudioFile::default defaulting to 0 volume

melody-rs committed:
Bump rodio version (fixes sound test crash?)

melody-rs committed:
Fix compiler warnings

melody-rs committed:
Merge branch 'dev' into event-editor
